### PR TITLE
Pop downscaling

### PR DESCRIPTION
**Pull request by knoiva-indecon**
Merging in population downscaling functionality to state level

### CODE REVIEW

File: FrEDI/R/run_fredi_sv.R [hunk 1]
```diff
@@ -2,47 +2,49 @@
...
+#' `run_fredi_sv` allows users to project annual average climate change impacts throughout the 21st century (2010-2100) for socially vulnerable (SV) populations for available sectors. Users can run [FrEDI::run_fredi_sv()] for individual sectors to generate annual physical impacts for SV populations. [FrEDI::run_fredi_sv()] can be run with default population and climate (temperature and sea level rise trajectories) or provide custom trajectories. The output of [FrEDI::run_fredi_sv()] is an R data frame object containing annual average physical impacts at five-year increments for the period 2010 to 2100. Users have the option to write outputs to Excel files that provide additional visualization of SV outputs.
```
Comment (emcduffie) [inline FrEDI/R/run_fredi_sv.R:5]:
```suggestion
#' `run_fredi_sv` allows users to project annual average climate change impacts throughout the 21st century (2010-2100) for socially vulnerable (SV) populations for available sectors. Users can run [FrEDI::run_fredi_sv()] for individual sectors to generate annual physical impacts for SV populations. [FrEDI::run_fredi_sv()] can be run with default population and climate (temperature and sea level rise trajectories) or using custom trajectories. The output of [FrEDI::run_fredi_sv()] is an R data frame object containing annual average physical impacts at five-year increments for the period 2010 to 2100. The basic structure, specific methodology, and underlying data supporting FrEDI-SV are derived from EPA’s independently peer-reviewed September 2021 report, [Climate Change and Social Vulnerability in the United States: A Focus on Six Impacts](https://epa.gov/cira/social-vulnerability-report)
```

File: FrEDI/R/utils_import_inputs.R [hunk 2]
```diff
@@ -29,81 +77,456 @@ fun_tryInput <- function(
...
-  }
+  } ### End if(!(filename |> is.null()))
 
   ###### Return ######
   return(return_list)
 }
 
 
-###### rename_inputs ######
-### Created 2021.02.08. Last updated 2021.02.08
```
Comment (emcduffie) [inline FrEDI/R/utils_import_inputs.R:57]: Up to you whether you want to keep these 'Last Updated' dates throughout all the code in the R package. If so, I would recommend going through and updating the relevant dates before merging.

File: FrEDI/testing/utils_create_report_figures.R [hunk 3]
```diff
@@ -810,9 +882,69 @@ plot_DoW_by_modelYear <- function(
...
+  return(df_types)
+} ### fun_create_df_types
+
+
 ### Plot DOW
 plot_DoW <- function(
-    df0,         ### Data (e.g., output from sum_impactsByDegree)
+    df0,     ### Data (e.g., output from sum_impactsByDegree)
     types0   = c("GCM", "SLR"), ### Model type: GCM or SLR
     years0   = c(2010, 2090),
```
Comment (emcduffie) [inline FrEDI/testing/utils_create_report_figures.R:949]: plot through 2100?

File: FrEDI/testing/utils_create_report_figures.R [hunk 4]
```diff
@@ -906,77 +1024,77 @@ plot_DoW_by_sector <- function(
       theme      = NULL
     )
 ){
+  ### Values & Conditions
+  years0     <- c(2010, 2090)
```
Comment (emcduffie) [inline FrEDI/testing/utils_create_report_figures.R:1028]: plot 2100 results instead of 2090?

**Review by emcduffie — CHANGES_REQUESTED** [review]
Looks great, just a few minor suggested text edits!

**Comment by knoiva-indecon** [comment]
Awesome, thanks for the review! I’ll make those edits later today or tomorrow.
________________________________

File: FrEDI/R/import_inputs.R [hunk 5]
```diff
@@ -1,40 +1,56 @@
...
+#'    * GDP values must be greater than or equal to zero.
+#'    * The GDP scenario must have at least one non-missing value in 2010 or earlier and at least one non-missing value in or after the final analysis year (as specified by the [FrEDI::run_fredi()] `maxYear` argument).
+#' * __Population Scenario.__ The file specified by `popfile` population must have at least two columns -- `"year"` and `"pop"` -- respectively containing the years associated with the population estimates and the population estimates. requires state-level population values. Whether a third column is required depends on the geographical scale of the data, as specified by the `popArea` argument.
+#'    * If `popArea = "state"`, the third column should be called `"state"` and contain the name of the state associated with each population estimate in a given year. If `popArea = "regional"`, the third column should be called `"region"` and contain the NCA Region label associated with each population estimate in a given year. If `popArea = "national"` or `popArea = "area"`, only the `year` and `pop` columns are required.
```
[the remark this answers is not in the document]
Comment (knoiva-indecon) [reply FrEDI/R/import_inputs.R]: Removed option `popArea="area"` from options in merge #174. Basically, it was an option if the user happened to have CONUS populations and estimates for Alaska and Hawaii. The function is still set up to handle `popArea="area"`, but it seems like a confusing option for a casual user, and it seems like users would be less likely to have inputs at that scale. Let me know if you think differently.

File: FrEDI/R/data.R
[the remark this answers is not in the document]
Comment (knoiva-indecon) [reply FrEDI/R/data.R:1]: Made edits to line 57 and clarified using GCAM scenarios as inputs in merge #174. Will get back to you on default GDP scenarios

File: FrEDI/R/import_inputs.R [hunk 6]
```diff
@@ -1,40 +1,56 @@
...
-#'    * Population inputs must have at least one non-missing value in 2010 or earlier and at least one non-missing value in or after the final analysis year (as specified by the [FrEDI::run_fredi()] `maxYear` argument).
+#' [FrEDI::run_fredi()] requires a population scenario at the state level. If the population scenario is supplied to [FrEDI::import_inputs()] at a geographical scale above the state level (i.e., if `popArea = "national"`, `popArea = "area"`, or `popArea = "regional"`), [FrEDI::import_inputs()] will calculate state-level estimates from the provided data:
 #'
+#' * If `popArea = "national"`, [FrEDI::import_inputs()] will use historical U.S. Census data for the period 2010--2023 to allocate total national population to CONUS and non-CONUS regions (i.e., Alaska and Hawaii); U.S. Census values from 2023 are applied in allocations for years after 2023.
+#' * If `popArea = "national"` or or `popArea = "area"`, [FrEDI::import_inputs()] uses ICLUS data and projections for the period 2010--2100 to allocate CONUS population to specific NCA regions; ICLUS values from 2100 are applied in allocations for years after 2100.
```
[the remark this answers is not in the document]
Comment (knoiva-indecon) [reply FrEDI/R/import_inputs.R]: Fixed in merge #174

File: FrEDI/R/run_fredi.R [hunk 7]
```diff
@@ -4,109 +4,108 @@
...
+#' If `inputsList = NULL`, [FrEDI::run_fredi()] uses defaults for temperature, SLR, GDP, and population. Otherwise, [FrEDI::run_fredi()] looks for a list object passed to the argument `inputsList`. Within that list, [FrEDI::run_fredi()] looks for list elements `tempInput`, `slrInput`, `gdpInput`, and `popInput` containing data frames with custom scenarios for temperature, GMSL, GDP, and regional population, respectively. [FrEDI::run_fredi()] will default back to the default scenarios for any list elements that empty or `NULL` (in other words, running `run_fredi(inputsList = list())` returns the same outputs as running [FrEDI::run_fredi()]).
```
[the remark this answers is not in the document]
Comment (knoiva-indecon) [reply FrEDI/R/run_fredi.R]: in merge #175

File: FrEDI/R/utils_import_inputs.R (hunk 2 above)
Comment (knoiva-indecon) [reply FrEDI/R/utils_import_inputs.R:57]: Removed from files in merge #175

File: FrEDI/testing/utils_create_report_figures.R (hunk 4 above)
Comment (knoiva-indecon) [reply FrEDI/testing/utils_create_report_figures.R:1028]: This is for impact years, and we have impact year estimates for 2090, so didn't change

File: FrEDI/testing/utils_create_report_figures.R (hunk 3 above)
Comment (knoiva-indecon) [reply FrEDI/testing/utils_create_report_figures.R:949]: This is for impact years, and we have impact year estimates for 2090, so didn't change

**Comment by knoiva-indecon** [comment]
@emcduffie : Merge #176 tries to meet your offline request to add some flexibility to input column names:

Added flexibility to column names: Now, if the driver column is not found, the import_inputs helper function will try to string match the input type against column names. If it finds a match: 

> - If a temperature input, it will look for matches to strings "conus" and "global". If conus matches, it will use that column and set tempType to "conus". Else if global matches, it will use that column and set tempType to "global". Otherwise, it will use the first matched column. 
> - If a population input, it will look for matches to strings "state", "region", "conus", and "global". It will check for a match and use the first associated column, and set popArea accordingly. E.g., if "reg_pop" is present, but "state_pop" isn't, "reg_pop" will be used and the popArea set to "regional".
> - Otherwise, the first matched column will be used as the value column.

The function will then rename the matched column to the expected column. No flexibility enhancements were made for ID columns.

**Review by emcduffie — APPROVED** [review]
Thanks for addressing the changes. Please go ahead with merging into main.

**Pull request merged**